diff --git a/MODULE.bazel b/MODULE.bazel
index 0ee5dc672b..f8e71cbf6c 100644
--- a/MODULE.bazel
+++ b/MODULE.bazel
@@ -24,7 +24,7 @@ module(
 
 # Upgrade version of rules_swift to avoid an error in BCR CI of not
 # finding switfc.exe on Windows with bazel 7.
-bazel_dep(name = "rules_swift", version = "2.5.0")
+bazel_dep(name = "rules_swift", version = "3.5.0")
 
 # -- Abseil library.
 # Pinning version to be consistent with third_party/abseil-cpp
diff --git a/bazel/grpc_build_system.bzl b/bazel/grpc_build_system.bzl
index 00c86b82a0..774950a054 100644
--- a/bazel/grpc_build_system.bzl
+++ b/bazel/grpc_build_system.bzl
@@ -31,7 +31,7 @@ load("@build_bazel_apple_support//rules:universal_binary.bzl", "universal_binary
 load("@build_bazel_rules_apple//apple:ios.bzl", "ios_unit_test")
 load("@build_bazel_rules_apple//apple/testing/default_runner:ios_test_runner.bzl", "ios_test_runner")
 load("@com_google_protobuf//bazel:cc_proto_library.bzl", "cc_proto_library")
-load("@com_google_protobuf//bazel:upb_proto_library.bzl", "upb_proto_library", "upb_proto_reflection_library")
+load("@com_google_protobuf//upb/bazel:upb_proto_library.bzl", "upb_proto_library", "upb_proto_reflection_library")
 load("@rules_cc//cc:cc_binary.bzl", "cc_binary")
 load("@rules_cc//cc:cc_library.bzl", "cc_library")
 load("@rules_cc//cc:cc_test.bzl", "cc_test")
diff --git a/src/core/lib/event_engine/ares_resolver.cc b/src/core/lib/event_engine/ares_resolver.cc
index 2d80ce22a4..3173e7e017 100644
--- a/src/core/lib/event_engine/ares_resolver.cc
+++ b/src/core/lib/event_engine/ares_resolver.cc
@@ -648,7 +648,7 @@ void AresResolver::OnAresBackupPollAlarm() {
 
 void AresResolver::OnHostbynameDoneLocked(void* arg, int status,
                                           int /*timeouts*/,
-                                          struct hostent* hostent) {
+                                          const struct hostent* hostent) {
   auto* hostname_qa = static_cast<HostnameQueryArg*>(arg);
   GRPC_CHECK_GT(hostname_qa->pending_requests--, 0);
   auto* ares_resolver = hostname_qa->ares_resolver;
@@ -744,7 +744,7 @@ void AresResolver::OnHostbynameDoneLocked(void* arg, int status,
 }
 
 void AresResolver::OnSRVQueryDoneLocked(void* arg, int status, int /*timeouts*/,
-                                        unsigned char* abuf, int alen) {
+                                        const unsigned char* abuf, int alen) {
   std::unique_ptr<QueryArg> qa(static_cast<QueryArg*>(arg));
   auto* ares_resolver = qa->ares_resolver;
   auto nh = ares_resolver->callback_map_.extract(qa->callback_map_id);
@@ -808,7 +808,7 @@ void AresResolver::OnSRVQueryDoneLocked(void* arg, int status, int /*timeouts*/,
 }
 
 void AresResolver::OnTXTDoneLocked(void* arg, int status, int /*timeouts*/,
-                                   unsigned char* buf, int len) {
+                                   const unsigned char* buf, int len) {
   std::unique_ptr<QueryArg> qa(static_cast<QueryArg*>(arg));
   auto* ares_resolver = qa->ares_resolver;
   auto nh = ares_resolver->callback_map_.extract(qa->callback_map_id);
diff --git a/src/core/lib/event_engine/ares_resolver.h b/src/core/lib/event_engine/ares_resolver.h
index ce258dade9..66f078cee9 100644
--- a/src/core/lib/event_engine/ares_resolver.h
+++ b/src/core/lib/event_engine/ares_resolver.h
@@ -132,13 +132,13 @@ class AresResolver : public RefCountedDNSResolverInterface {
   // These callbacks are invoked from the c-ares library, so disable thread
   // safety analysis. We are guaranteed to be holding mutex_.
   static void OnHostbynameDoneLocked(void* arg, int status, int /*timeouts*/,
-                                     struct hostent* hostent)
+                                     const struct hostent* hostent)
       ABSL_NO_THREAD_SAFETY_ANALYSIS;
   static void OnSRVQueryDoneLocked(void* arg, int status, int /*timeouts*/,
-                                   unsigned char* abuf,
+                                   const unsigned char* abuf,
                                    int alen) ABSL_NO_THREAD_SAFETY_ANALYSIS;
   static void OnTXTDoneLocked(void* arg, int status, int /*timeouts*/,
-                              unsigned char* buf,
+                              const unsigned char* buf,
                               int len) ABSL_NO_THREAD_SAFETY_ANALYSIS;
 #ifdef GRPC_ENABLE_FORK_SUPPORT
   // Is executed on fork before the poller is restarted. Cleans up the resources
diff --git a/src/core/resolver/dns/c_ares/grpc_ares_wrapper.cc b/src/core/resolver/dns/c_ares/grpc_ares_wrapper.cc
index 7fda1eb2b6..b2f3c4a4ba 100644
--- a/src/core/resolver/dns/c_ares/grpc_ares_wrapper.cc
+++ b/src/core/resolver/dns/c_ares/grpc_ares_wrapper.cc
@@ -676,7 +676,7 @@ static void destroy_hostbyname_request_locked(grpc_ares_hostbyname_request* hr)
 }
 
 static void on_hostbyname_done_locked(void* arg, int status, int /*timeouts*/,
-                                      struct hostent* hostent)
+                                      const struct hostent* hostent)
     ABSL_NO_THREAD_SAFETY_ANALYSIS {
   // This callback is invoked from the c-ares library, so disable thread safety
   // analysis. Note that we are guaranteed to be holding r->mu, though.
@@ -750,7 +750,7 @@ static void on_hostbyname_done_locked(void* arg, int status, int /*timeouts*/,
 }
 
 static void on_srv_query_done_locked(void* arg, int status, int /*timeouts*/,
-                                     unsigned char* abuf,
+                                     const unsigned char* abuf,
                                      int alen) ABSL_NO_THREAD_SAFETY_ANALYSIS {
   // This callback is invoked from the c-ares library, so disable thread safety
   // analysis. Note that we are guaranteed to be holding r->mu, though.
@@ -800,7 +800,7 @@ static void on_srv_query_done_locked(void* arg, int status, int /*timeouts*/,
 static const char g_service_config_attribute_prefix[] = "grpc_config=";
 
 static void on_txt_done_locked(void* arg, int status, int /*timeouts*/,
-                               unsigned char* buf,
+                               const unsigned char* buf,
                                int len) ABSL_NO_THREAD_SAFETY_ANALYSIS {
   // This callback is invoked from the c-ares library, so disable thread safety
   // analysis. Note that we are guaranteed to be holding r->mu, though.